# Schema Validation — Olist E-Commerce Dataset

This notebook validates the structure and key constraints of the Olist dataset before setting up the relational database.

The main focus is on:
- identifying candidate primary keys
- validating uniqueness
- checking null values
- understanding identifiers and relationships

## 1. Customers

### 1.1 Primary Key Validation — `customer_id`

#### Question

Can `customer_id` serve as the primary key of the `customers` table?

To answer this, we check:
- total number of records
- number of unique `customer_id` values
- number of null `customer_id` values

In [4]:
import pandas as pd

customers_df = pd.read_csv("../data/olist_customers_dataset.csv")

customers_df.head(10)

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP
5,879864dab9bc3047522c92c82e1212b8,4c93744516667ad3b8f1fb645a3116a4,89254,jaragua do sul,SC
6,fd826e7cf63160e536e0908c76c3f441,addec96d2e059c80c30fe6871d30d177,4534,sao paulo,SP
7,5e274e7a0c3809e14aba7ad5aae0d407,57b2a98a409812fe9618067b6b8ebe4f,35182,timoteo,MG
8,5adf08e34b2e993982a47070956c5c65,1175e95fb47ddff9de6b2b06188f7e0d,81560,curitiba,PR
9,4b7139f34592b3a31687243a302fa75b,9afe194fb833f79e300e37e580171f22,30575,belo horizonte,MG


In [5]:
customers_df.shape[0]

99441

In [6]:
customers_df["customer_id"].nunique()

99441

In [7]:
customers_df["customer_id"].isna().sum()

np.int64(0)

#### Insight

`customer_id` contains 99,441 unique and non-null values across 99,441 records. Therefore, it satisfies the uniqueness and non-null requirements for a primary key in the `customers` table.

### 1.2 Customer Identifier Validation — `customer_unique_id`

#### Question

Is `customer_unique_id` unique for each record in the `customers` table, or can the same customer appear in multiple records?

To answer this, we compare:
- total number of records
- number of unique `customer_unique_id` values
- number of null `customer_unique_id` values

In [8]:
customers_df.shape[0]

99441

In [9]:
customers_df["customer_unique_id"].nunique()

96096

In [10]:
customers_df["customer_unique_id"].isna().sum()

np.int64(0)

#### Insight

`customer_id` is unique for every record, while `customer_unique_id` is not unique: there are 99,441 records but only 96,096 unique `customer_unique_id` values.

This suggests that `customer_id` identifies a customer record, whereas `customer_unique_id` represents a customer at the business level and may appear in multiple records.

The relationship between repeated `customer_unique_id` values and multiple orders will be validated later using the `orders` table.

## 2. Orders

In [11]:
orders_df = pd.read_csv("../data/olist_orders_dataset.csv")

### 2.1 Primary Key Validation — `order_id`

#### Question

Can `order_id` serve as the primary key of the `orders` table?

To answer this, we check:
- total number of records
- number of unique `order_id` values
- number of null `order_id` values

In [15]:
orders_df.head(10)

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00
5,a4591c265e18cb1dcee52889e2d8acc3,503740e9ca751ccdda7ba28e9ab8f608,delivered,2017-07-09 21:57:05,2017-07-09 22:10:13,2017-07-11 14:58:04,2017-07-26 10:57:55,2017-08-01 00:00:00
6,136cce7faa42fdb2cefd53fdc79a6098,ed0271e0b7da060a393796590e7b737a,invoiced,2017-04-11 12:22:08,2017-04-13 13:25:17,NaN,NaN,2017-05-09 00:00:00
7,6514b8ad8028c9f2cc2374ded245783f,9bdf08b4b3b52b5526ff42d37d47f222,delivered,2017-05-16 13:10:30,2017-05-16 13:22:11,2017-05-22 10:07:46,2017-05-26 12:55:51,2017-06-07 00:00:00
8,76c6e866289321a7c93b82b54852dc33,f54a9f0e6b351c431402b8461ea51999,delivered,2017-01-23 18:29:09,2017-01-25 02:50:47,2017-01-26 14:16:31,2017-02-02 14:08:10,2017-03-06 00:00:00
9,e69bfb5eb88e0ed6a785585b27e16dbf,31ad1d1b63eb9962463f764d4e6e0c9d,delivered,2017-07-29 11:55:02,2017-07-29 12:05:32,2017-08-10 19:45:24,2017-08-16 17:14:30,2017-08-23 00:00:00


In [16]:
orders_df.shape[0]

99441

In [17]:
orders_df["order_id"].nunique()

99441

In [18]:
orders_df["order_id"].isna().sum()

np.int64(0)

### Insight

`order_id` contains 99,441 unique and non-null values across 99,441
records. Therefore, it satisfies the uniqueness and non-null requirements
for a primary key in the `orders` table.

### 2.2 Foreign Key Validation — `customer_id`

#### Question

Does `orders.customer_id` correctly reference `customers.customer_id`?

To answer this, we need to check:
- whether `orders.customer_id` contains null values
- whether every `customer_id` in `orders` exists in the `customers` table

In [19]:
orders_df["customer_id"].isna().sum()

np.int64(0)

In [28]:
(~orders_df["customer_id"].isin(customers_df["customer_id"])).sum()

np.int64(0)

#### Insight

All `customer_id` values in `orders` exist in `customers`, so `customer_id` is a valid foreign key.